### PR TITLE
Support intra-site links to source content files

### PR DESCRIPTION
**Pull request by wombelix**
As stated in the README, the Plugin doesn't support internal linking like:

reStructuredText

```
`a link relative to the current file <{filename}../article2.md>`_
`a link relative to the content root <{filename}/article2.md>`_
```

Markdown

```
[a link relative to the current file]({filename}category/article1.rst)
[a link relative to the content root]({filename}/category/article1.rst)
```

Due to the fact that the PDF Generation is based on the Article / Page Source and not the generated HTML Content, no conversion of the Placeholder like `{filename}` is done prior the Output.

It seems like that implementing a own version of `Content._get_intrasite_link_regex` with adjusted regular expressions and using `Content._link_replacer` could work.

RegEx to detect links in reStructuredText:
```
(?P<markup>)(?P<quote>)(?P<path>(\:?)[{|](?P<what>.*?)[|}](?P<value>.*?)(?=[>\n]))
```

Test Input:
```
.. _@testlink: {filename}../article2.md

`a link relative to the content root <{filename}/article2.md>`_

`a link relative to the content root <{filename}/article2.md>`_

.. _@wombelix: {filename}../article2.md

.. _@wombelix: {filename}../article2.md

`a link relative to the content root <{filename}/article2.md>`_

`a link relative to the content root <{filename}/article2.md>`_

```

@justinmayer maybe you have a moment to let me know if the approach make sense in your opinion. 
Technically it worked quite well in some tests today and produced the expected output, but i'm not that familiar with the Pelican code base yet, there might be a better way implementing this?

The relevant Code:
[pdf.py / R97-R104](https://github.com/pelican-plugins/pdf/compare/main...wombelix:internal_linking?expand=1#diff-f99d6ac7c1601521e4267e0f30ec0d51c29a9d1d4dbedd49f8ae2c9b7792aa81R97-R104)
[pdf.py / R59-62](https://github.com/pelican-plugins/pdf/compare/main...wombelix:internal_linking?expand=1#diff-f99d6ac7c1601521e4267e0f30ec0d51c29a9d1d4dbedd49f8ae2c9b7792aa81R59-R62)

### CODE REVIEW

**Comment by justinmayer** [comment]
@avaris: If you have a moment, could you have a look and let @wombelix know what you think?

**Comment by avaris** [comment]
> It seems like that implementing a own version of Content._get_intrasite_link_regex with adjusted regular expressions and using Content._link_replacer could work.

That sounds correct. But regex maybe a bit broad? If I'm not mistaken, you're only looking for `{filename}...` pattern, not necessarily links.

**Comment by wombelix** [comment]
> > It seems like that implementing a own version of Content._get_intrasite_link_regex with adjusted regular expressions and using Content._link_replacer could work.
> 
> That sounds correct. But regex maybe a bit broad? If I'm not mistaken, you're only looking for `{filename}...` pattern, not necessarily links.

Thanks for your Feedback, i was thinking about and checked again the way how it's implemented in Pelican Core.
If i wanna use `Content._link_replacer`, it seems regex is the way to go, because I have to pass a map were at least "path", "what" and "value" are given.

Also a link might be relative (e.g. {filename}../myfolder/myfile.rst) so only search for {filename} and for example replace it with the Siteurl might not sufficient?

Don't get me wrong, I'm not a big regex fan and it took me quite a while to came up with this approach, so if there is an easier solution without regex possible, i couldn't be happier :) 

I just want to ensure to implement it "right", even though I will first focus on RST and take a look at Markdown later.

**Comment by avaris** [comment]
Sorry, bad wording on my part. Regex is the way to go and your approach is correct. I meant the regex pattern you're using is not limited to links. You're looking for all `{filename}...` type patterns in the content. They might appear in the text without being actual links (consider what I'm writing right now, you'll get a match for the previous sentence). But given that you're working with source material and the multitude of ways links can happen in RST or MD, the regex that accounts for all of that would get pretty complex and I suppose current way is a good compromise.

**Comment by wombelix** [comment]
> Sorry, bad wording on my part. Regex is the way to go and your approach is correct. I meant the regex pattern you're using is not limited to links. You're looking for all `{filename}...` type patterns in the content. They might appear in the text without being actual links (consider what I'm writing right now, you'll get a match for the previous sentence). But given that you're working with source material and the multitude of ways links can happen in RST or MD, the regex that accounts for all of that would get pretty complex and I suppose current way is a good compromise.

No worries and thanks for the explanation, i was indeed looking for something that's not limited to {filename} and would also cover other internal links like {author}, {category}, {index}, {tag}, {static}.

Agree, might get a real regex monster to have all RST and MD edge cases covered, so it seems like a good starting point.

I will try to implement some tests and push a new commit within the next days.

And in case someone else except me is actually using the Plugin and has some trouble due to false positives, we can still try to improve the regex :)

**Comment by wombelix** [comment]
OK so I tested the above regex successfully with following Markdown Links:

```
[a link relative to the current file]({filename}category/article123.rst)
[a link relative to the content root]({filename}/category/article_123.rst)

[arbitrary case-insensitive reference text]: {filename}category/article456.rst
[1]: {filename}../category/article789.rst

[link text itself]: {filename}category/article_8-9-0.rst 
```

Code adjusted to support Markdown as well, new commit pushed for Review and Feedback.
No Unittest written yet (mandatory for a first release?)

Manual tests - focused on RST - were successful and covered all types of internal links, Example PDF attached.
[test_pdf_rst.pdf](https://github.com/pelican-plugins/pdf/files/6126313/test_pdf_rst.pdf)

**Comment by justinmayer** [comment]
@avaris: Do you perhaps have a moment to take another look at the latest iteration here?

**Comment by justinmayer** [comment]
@wombelix: Many thanks for your work on this! Would you be willing to add one or more test(s) for this linking-to-internal-content support? By including tests along with new functionality, we increase the likelihood that it won't be broken by future contributions. 😁

**Comment by wombelix** [comment]
@justinmayer I didn't had any issues yet by using the new internal linking so I think that can be considered to be merged. I'm running in some issues with the pre-commit.ci hooks, they seem to rewrite an import statement in the unittest and then the github action fails. Local pre-commit as well manual lint check are fine, not sure how to handle that, maybe you have an idea?

**Comment by justinmayer** [comment]
Hi Dominik. I'm sorry about the Pre-commit CI behavior — I added that as an experiment but am not yet convinced it is worth it. In any case, I updated the linter versions in the `main` branch, so if you rebase _your_ commits in this PR (dropping the automated Pre-commit CI commits) on current `main`, I believe the linters will be pacified. Will you please let me know once you've done that?

**Comment by wombelix** [comment]
> Hi Dominik. I'm sorry about the Pre-commit CI behavior — I added that as an experiment but am not yet convinced it is worth it. In any case, I updated the linter versions in the `main` branch, so if you rebase _your_ commits in this PR (dropping the automated Pre-commit CI commits) on current `main`, I believe the linters will be pacified. Will you please let me know once you've done that?

Hey Justin, thank you, I still struggling to pass the pre-commit check. As suggested I rebased my changes on top of the latest version of the main branch, everything seem to be fine, black/lint/isort are all good when i run them locally, but as soon pushed to github the check added by pre-commit is going to fail. Did I made something wrong? Any suggestion how I can fix that? Thanks in advance for your support.

**Comment by justinmayer** [comment]
> I am still struggling to pass the pre-commit check. As suggested I rebased my changes on top of the latest version of the main branch, everything seem to be fine, black/lint/isort are all good when i run them locally, but as soon pushed to github the check added by pre-commit is going to fail. Did I made something wrong? Any suggestion how I can fix that?

You didn’t do anything wrong. I have seen this import-ordering inconsistency before, and it is maddening. I added a commit in `main` that I believe should resolve that local-versus-CI import linter inconsistency.

If you don’t mind, would you please [1] rebase on current `main`, [2] drop the spurious automatic commit by pre-commit.ci (again), and [3] add a `RELEASE.md` file in the project root containing the following?

```markdown
Release type: patch

Support intra-site links to source content files
```

**Comment by wombelix** [comment]
> If you don’t mind, would you please [1] rebase on current `main`, [2] drop the spurious automatic commit by pre-commit.ci (again), and [3] add a `RELEASE.md` file in the project root containing the following?

Sure, thank you for fixing the pre-commit issue, looks like all went fine this time :)

**Comment by wombelix** [comment]
@justinmayer I just saw that there was some duplicate code that was worth refactoring. Ready to be reviewed again if you have some time.

**Review by avaris — APPROVED** [review]
:+1:

**Comment by justinmayer** [comment]
The following stanza in the PDF plugin's test file appears to explicitly refer to the _lack_ of the feature for which this PR adds support. Should this stanza be modified to account for this new feature?

https://github.com/pelican-plugins/pdf/blob/main/pelican/plugins/pdf/test_pdf.py#L32-L40

**Comment by wombelix** [comment]
> The following stanza in the PDF plugin's test file appears to explicitly refer to the _lack_ of the feature for which this PR adds support. Should this stanza be modified to account for this new feature?
> 
> https://github.com/pelican-plugins/pdf/blob/main/pelican/plugins/pdf/test_pdf.py#L32-L40

Thank you, good catch, I removed the try/catch block and adjusted the test_data a little

File: pelican/plugins/pdf/test_pdf.py [hunk 1]
```diff
@@ -29,15 +28,8 @@ def setUp(self, override=None):
...
-            pelican.run()
-        except ValueError:
-            logging.warn(
-                "Relative links in the form of "
-                + "|filename|images/test.png are not yet handled by "
-                + " the pdf generator"
-            )
-            pass
+        pelican.run()
+        pass
```
Comment (justinmayer) [inline pelican/plugins/pdf/test_pdf.py]: What purpose does `pass` serve here?

**Review by justinmayer — COMMENTED** [review]
Thank you for the follow-up improvements. Would you let me know about the question I raised about the `pass` directive?

**Comment by wombelix** [comment]
> Thank you for the follow-up improvements. Would you let me know about the question I raised about the `pass` directive?

Indeed, doesn't make much sense anymore, removed both and pushed a new commit

**Review by justinmayer — APPROVED** [review]
Many thanks to @wombelix for patiently working on this feature, and thanks to @avaris for helping review it. 🥇
